### DISCUSSION TITLE
nixos-with-flakes/other-useful-tips

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#101] [Announcements]
# nixos-with-flakes/other-useful-tips

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/nixos-with-flakes/other-useful-tips

**Comment by maribox** [comment]
Is it really a good idea to let all system settings be changed with user-level privileges?

**Reply by ryan4yin** [reply]
Remember the frequently used command `sudo nixos-rebuild switch` earlier in this book? These changes must be deployed in privileged mode to take effect, and you can always check what has been changed via `git` before deploying your nixos config.

**Reply by maribox** [reply]
Yeah but these changes could also be tampered with if a program has user-level access, right? And even if we can see all changes since the last build correctly, the user might forget to check git before rebuilding... Would it be a valid alternative to just run git on the actual folder with root privileges?

**Reply by ryan4yin** [reply]
I think this is more insecure, because you usually want to use tools such as Neovim/VSCode to help write nix code, do code formatting, syntax checking, error prompts, etc., then you have to give these editors and all their plug-ins root permission, which is obviously riskier.

**Reply by maribox** [reply]
Understandable, thanks!

**Comment by vaibd** [comment]
What about hardware-configuration.nix file? In the file itself it says may be overwritten. So, is it okay to sym link it?

**Reply by PhilVoel** [reply]
It's a nix file like any other and you can edit it all you want. But it was created based on yor system when you set up NixOS so it has settings that work with the system.
So without backup and knowing what you're doing you can seriously brick your system if you change it.
That's why a beginner **shouldn't** edit it but you definitely **can**.
But as I said, apart from containing some critical hardware-specific configs, it is not special. You could even move its contents into your main config file and delete hardware.nix altogether

**Comment by hackerncoder** [comment]
> \# Wiping history won't garbage collect the unused packages, you need to run the gc command manually as root:
> sudo nix-collect-garbage --delete-old

`--delete-old` will delete all old profiles! The way it is written this is not clear.

**Comment by elvetemedve** [comment]
The section "**Viewing and Deleting Historical Data**" only deals with globally installed packages. I'm wondering how to do the same for **Home Manager** installed packages. I've tried this command below, but returned no results:
`nix profile history --profile /etc/profiles/per-user/$USER`

Though I found the list of generations by `ls -l ~/.local/state/nix/profiles` and the list of installed packages by `nix-store -q --requisites /etc/profiles/per-user/$USER`, it would be nice to use `nix profile` for that as well.

**Reply by sud0pacman** [reply]
I think you have the same issue, right ?

https://github.com/NixOS/nix/issues/8508

this issue solution is:
nix-collect-garbage --delete-old

**Comment by Sporarum** [comment]
For #managing-the-configuration-with-git
It's also possible to set the symlink in configuration.nix:
```nix
{ config, pkgs, ...}:
{
  // other options
  environment.etc.nixos.source = "/path/to/flake";
  // other options
};
```
IIRC the path has to be a string and not a path literal, something to do with path literals being added to the nix store
